### PR TITLE
[BEAM-8280][BEAM-8629] Make IOTypeHints immutable

### PR DESCRIPTION
**Pull request by udim**
Also adds an `origin` field that hopefully aids in debugging type check
errors.

This may break uses of `@apache_beam.decorators.with_{input,output}_types(...)` on builtins and Cython functions. For Cython, a possible solution is to use `# cython: binding=True`.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
run python 2 postcommit

**Comment by udim** [comment]
run python 3.7 postcommit

**Comment by udim** [comment]
run python 2 postcommit

**Comment by udim** [comment]
run python 3.7 postcommit

**Comment by udim** [comment]
R: @chadrik

File: sdks/python/apache_beam/typehints/decorators.py [hunk 1]
```diff
@@ -305,7 +326,7 @@ def has_simple_output_type(self):
     return (self.output_types and len(self.output_types[0]) == 1 and
             not self.output_types[1])
 
-  def strip_iterable(self):
+  def strip_iterable(self):  # type: (...) -> IOTypeHints
```
Comment (chadrik) [inline sdks/python/apache_beam/typehints/decorators.py]: Type comment should go after the function.  No need for `...` unless you are skipping the definition of an argument (the type of `self` is implicit). 

```suggestion
  def strip_iterable(self):
    # type: () -> IOTypeHints
```

File: sdks/python/apache_beam/typehints/decorators.py [hunk 2]
```diff
@@ -283,16 +301,19 @@ def from_callable(fn):
...
 
-  def set_input_types(self, *args, **kwargs):
-    self.input_types = args, kwargs
+  def with_input_types(self, *args, **kwargs):  # type: (...) -> IOTypeHints
+    return self._replace(input_types=(args, kwargs),
+                         origin=self._make_traceback(self))
 
-  def set_output_types(self, *args, **kwargs):
-    self.output_types = args, kwargs
+  def with_output_types(self, *args, **kwargs):  # type: (...) -> IOTypeHints
```
Comment (udim) [inline sdks/python/apache_beam/typehints/decorators.py]:
```suggestion
  def with_output_types(self, *args, **kwargs):
    # type: (...) -> IOTypeHints
```

File: sdks/python/apache_beam/typehints/decorators.py [hunk 3]
```diff
@@ -283,16 +301,19 @@ def from_callable(fn):
       output_args.append(typehints.Any)
 
     return IOTypeHints(input_types=(tuple(input_args), input_kwargs),
-                       output_types=(tuple(output_args), {}))
+                       output_types=(tuple(output_args), {}),
+                       origin=cls._make_traceback(None))
 
-  def set_input_types(self, *args, **kwargs):
-    self.input_types = args, kwargs
+  def with_input_types(self, *args, **kwargs):  # type: (...) -> IOTypeHints
```
Comment (udim) [inline sdks/python/apache_beam/typehints/decorators.py]:
```suggestion
  def with_input_types(self, *args, **kwargs):
    # type: (...) -> IOTypeHints
```

**Comment by udim** [comment]
Postcommit results (on previous commit) seem okay. The failure in Python37_PC seems transient.

**Comment by udim** [comment]
@chadrik Any more comments? Does this LG?

File: sdks/python/apache_beam/typehints/decorators.py [hunk 4]
```diff
@@ -217,7 +220,10 @@ def get_signature(func):
   return signature
 
 
-class IOTypeHints(object):
+class IOTypeHints(NamedTuple('IOTypeHints', [
+    ('input_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
+    ('output_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
+    ('origin', List[str])])):
```
Comment (chadrik) [inline sdks/python/apache_beam/typehints/decorators.py]: My guess is that this is going to trip up mypy, but we can defer that to a future mypy PR if you'd like.  This is the pattern that I use for these cases:

```python
_IOTypeHints = NamedTuple('_IOTypeHints', [
    ('input_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
    ('output_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
    ('origin', List[str])
])

class IOTypeHints(_IOTypeHints):
```

File: sdks/python/apache_beam/typehints/decorators.py [hunk 5]
```diff
@@ -228,18 +234,30 @@ class IOTypeHints(object):
...
+    tb_lines = 'TH>' + ''.join(tb[:-num_frames_skip]).replace('\n', '\nTH>')
+
+    res = [tb_lines + '\nbased on: ' + str(base)]
+    if base is not None:
+      res += base.origin
+    return res
+
+  @classmethod
+  def empty(cls):
+    return IOTypeHints(None, None, [])
```
Comment (chadrik) [inline sdks/python/apache_beam/typehints/decorators.py]: would be nice to add a type annotation to this:

```python
  def empty(cls):
    # type: () -> IOTypeHints
    return IOTypeHints(None, None, [])
```

**Comment by chadrik** [comment]
> @chadrik Any more comments? Does this LG?

Added two more minor notes.  Otherwise, LGTM.

File: sdks/python/apache_beam/typehints/decorators.py (hunk 4 above)
Comment (chadrik) [reply sdks/python/apache_beam/typehints/decorators.py]: The tradeoff is that this produces a sub-par repr for `IOTypeHints`:

```python
In [2]: from typing import *

In [3]: _IOTypeHints = NamedTuple('_IOTypeHints', [
   ...:     ('input_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
   ...:     ('output_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
   ...:     ('origin', List[str])
   ...: ])

In [4]: class IOTypeHints(_IOTypeHints):pass

In [5]: x = IOTypeHints(None, None, ['foo'])

In [6]: x
Out[6]: _IOTypeHints(input_types=None, output_types=None, origin=['foo'])
```

Doing a quick investigation to see if there's a hack fix for this.

Comment (chadrik) [reply sdks/python/apache_beam/typehints/decorators.py]: Something like this:

```python
    def __repr__(self):
      repr_fmt = '{typename}({repr_fmt})'.format(
            typename=type(self).__name__, 
            repr_fmt=', '.join('{name}=%r'.format(name=name) for name in self._fields)
      )
      return repr_fmt % self
```

Comment (udim) [reply sdks/python/apache_beam/typehints/decorators.py]: Trip up how?

I see this error in lint (should be fixed in next commit):
```
apache_beam/typehints/decorators.py:353: error: Value of type "Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]" is not indexable  [index]
```
This is the line:
```
output_type = self.output_types[0][0]
```
So it seems that mypy is correctly parsing these type hints.

File: sdks/python/apache_beam/typehints/decorators.py (hunk 5 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/decorators.py]: done

File: sdks/python/apache_beam/typehints/decorators.py (hunk 4 above)
Comment (chadrik) [reply sdks/python/apache_beam/typehints/decorators.py]: I assumed that mypy would need the intermediate class _or_ would want that class's name to match the name provided to `NamedTuple()`.  For example, both of these are invalid:

```python
class Foo(Generic[TypeVar('T')]):  # TypeVar needs to be defined above
    pass

T = TypeVar('T_with_suffix')  # TypeVar name needs to match variable name
class Bar(Generic[T]):
    pass
```

I just did my own isolated test of this and you're right, mypy has no issue with it, so it appears there's nothing special about the name given to `NamedTuple`, as far as mypy is concerned.

Thanks for taking the time to test that.  I was trying to give some quick feedback without getting too drawn in, but once I _did_ get drawn in, I should have gone back and checked my original assumption!

**Comment by chadrik** [comment]
looks good!

**Comment by udim** [comment]
Run Portable_Python PreCommit

**Comment by udim** [comment]
Run Portable_Python PreCommit

**Pull request merged**